File: frontend/src/lib/shared/mutators/positioning.ts
/**
 * Positioning Mutator
 * Provides offline-capable positioning that matches Rails positioning gem API
 *
 * Handles automatic position assignment with offline conflict resolution:
 * - New records get positioned at end of list (max position + 1)
 * - Position updates are cached offline and resolved server-side
 * - Supports scoped positioning (e.g., tasks within a job)
 */

import type { MutatorContext, MutatorFunction } from './base-mutator';

/**
 * Interface for records that support positioning
 */
export interface Positionable {
  position?: number;
  [key: string]: any; // Allow for scope fields like job_id, parent_id etc
}

/**
 * Configuration for positioning behavior
 */
export interface PositioningConfig {
  /** Position field name (defaults to 'position') */
  positionField?: string;
  /** Scope fields that group positioning (e.g., ['job_id'] means position is scoped within job) */
  scopeFields?: string[];
  /** Whether to allow manual position assignment (defaults to true) */
  allowManualPositioning?: boolean;
}

/**
 * Default positioning configuration
 */
const DEFAULT_POSITIONING_CONFIG: Required<PositioningConfig> = {
  positionField: 'position',
  scopeFields: [],
  allowManualPositioning: true,
};

/**
 * Cache for tracking next positions in offline scenarios
 * Key format: "tableName:scopeValue1:scopeValue2:..."
 */
const offlinePositionCache = new Map<string, number>();

/**
 * Generate cache key for position tracking
 */
function generatePositionCacheKey(tableName: string, data: any, scopeFields: string[]): string {
  const scopeValues = scopeFields.map((field) => data[field] || 'null').join(':');
  return scopeFields.length > 0 ? `${tableName}:${scopeValues}` : tableName;
}

/**
 * Get next available position for offline scenarios
 * Uses cached values to prevent conflicts when multiple records are created offline
 */
function getNextOfflinePosition(tableName: string, data: any, scopeFields: string[]): number {
  const cacheKey = generatePositionCacheKey(tableName, data, scopeFields);
  const currentMax = offlinePositionCache.get(cacheKey) || 0;
  const nextPosition = currentMax + 1;
  offlinePositionCache.set(cacheKey, nextPosition);
  return nextPosition;
}

/**
 * Clear position cache for a specific scope (useful when coming back online)
 */
export function clearPositionCache(
  tableName?: string,
  scopeData: any = {},
  scopeFields: string[] = []
): void {
  if (tableName === undefined) {
    // Clear all cache entries
    offlinePositionCache.clear();
  } else {
    const cacheKey = generatePositionCacheKey(tableName, scopeData, scopeFields);
    offlinePositionCache.delete(cacheKey);
  }
}

/**
 * Create positioning mutator for a specific table and configuration
 */
export function createPositioningMutator(
  tableName: string,
  config: PositioningConfig = {}
): MutatorFunction<Positionable> {
  const finalConfig = { ...DEFAULT_POSITIONING_CONFIG, ...config };

  return (data: Positionable, context?: MutatorContext): Positionable => {
    const { positionField, scopeFields, allowManualPositioning } = finalConfig;

    // Skip positioning if:
    // 1. This is an update and position isn't being changed
    // 2. Manual positioning is disabled and a position was provided
    if (context?.action === 'update' && !(positionField in data)) {
      return data;
    }

    if (!allowManualPositioning && data[positionField] !== undefined) {
      // Remove manually set position if not allowed and assign new position
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { [positionField]: _position, ...dataWithoutPosition } = data;
      const nextPosition = context?.offline
        ? getNextOfflinePosition(tableName, data, scopeFields)
        : Date.now(); // Use timestamp for online scenarios to avoid conflicts

      return {
        ...dataWithoutPosition,
        [positionField]: nextPosition,
      };
    }

    // If position is already set and manual positioning is allowed, keep it
    if (allowManualPositioning && data[positionField] !== undefined) {
      return data;
    }

    // For new records or updates without position, assign next position
    if (
      context?.action === 'create' ||
      (context?.action === 'update' && data[positionField] === undefined)
    ) {
      const nextPosition = context?.offline
        ? getNextOfflinePosition(tableName, data, scopeFields)
        : Date.now(); // Use timestamp for online scenarios to avoid conflicts

      return {
        ...data,
        [positionField]: nextPosition,
      };
    }

    return data;
  };
}

/**
 * Rails positioning gem compatible methods
 * These can be used in components for repositioning operations
 */
export class PositionManager {
  constructor(
    private tableName: string,
    private config: PositioningConfig = {}
  ) {}

  /**
   * Move record to specific position (Rails: move_to!)
   * This creates update data that will be handled by the positioning mutator
   */
  moveTo(newPosition: number): Partial<Positionable> {
    const positionField = this.config.positionField || 'position';
    return {
      [positionField]: newPosition,
    };
  }

  /**
   * Move record to top (Rails: move_to_top!)
   */
  moveToTop(): Partial<Positionable> {
    return this.moveTo(1);
  }

  /**
   * Move record to bottom (Rails: move_to_bottom!)
   * In offline scenarios, this uses a high timestamp
   */
  moveToBottom(): Partial<Positionable> {
    const positionField = this.config.positionField || 'position';
    return {
      [positionField]: Date.now() + 1000000, // Ensure it's higher than auto-assigned positions
    };
  }

  /**
   * Move record higher by one position (Rails: move_higher!)
   * Note: This requires knowing current position, typically used in UI with current data
   */
  moveHigher(currentPosition: number): Partial<Positionable> {
    return this.moveTo(Math.max(1, currentPosition - 1));
  }

  /**
   * Move record lower by one position (Rails: move_lower!)
   */
  moveLower(currentPosition: number): Partial<Positionable> {
    return this.moveTo(currentPosition + 1);
  }
}

/**
 * Common positioning mutators for typical use cases
 */

/**
 * Task positioning (scoped by job_id)
 */
export const taskPositioningMutator = createPositioningMutator('tasks', {
  scopeFields: ['job_id'],
  allowManualPositioning: true,
});

/**
 * Task positioning manager for UI operations
 */
export const TaskPositionManager = new PositionManager('tasks', {
  scopeFields: ['job_id'],
});

/**
 * Generic positioning (no scope)
 */
export const genericPositioningMutator = createPositioningMutator('generic');

/**
 * Generic positioning manager
 */
export const GenericPositionManager = new PositionManager('generic');

/**
 * Type helpers for positioning
 */
export type PositionUpdateData<T> = T & Partial<Positionable>;

File: lib/shared/mutators/positioning.rb
# frozen_string_literal: true

##
# Positioning Mutator
# Provides server-side positioning logic that integrates with Rails positioning gem
#
# This mutator:
# - Assigns positions for new records when not explicitly set
# - Validates position updates
# - Handles offline conflict resolution by normalizing positions
# - Integrates with existing Rails positioning gem functionality

module Shared
  module Mutators
    class Positioning
      # Configuration for positioning behavior
      PositioningConfig = Struct.new(
        :position_field,      # Position field name (defaults to :position)
        :scope_fields,        # Scope fields that group positioning (e.g., [:job_id])
        :allow_manual,        # Whether to allow manual position assignment
        keyword_init: true
      )

      # Default positioning configuration
      DEFAULT_CONFIG = PositioningConfig.new(
        position_field: :position,
        scope_fields: [],
        allow_manual: true
      ).freeze

      class << self
        ##
        # Apply positioning logic to record data
        #
        # @param record [ApplicationRecord] The record being mutated
        # @param data [Hash] The data being applied
        # @param context [Hash] Mutator context
        # @param config [PositioningConfig] Positioning configuration
        # @return [Hash] Modified data with positioning applied
        def apply(record, data, context, config = DEFAULT_CONFIG)
          position_field = config.position_field
          scope_fields = config.scope_fields

          # Skip positioning if:
          # 1. This is an update and position isn't being changed
          # 2. Manual positioning is disabled and a position was provided
          if context[:action] == :update && !data.key?(position_field)
            return data
          end

          unless config.allow_manual
            data = data.except(position_field) if data.key?(position_field)
          end

          # If position is set and manual positioning allowed, validate and keep it
          if config.allow_manual && data.key?(position_field)
            validate_position(record, data[position_field], data, scope_fields)
            return data
          end

          # For new records or updates that include position field (nil or undefined), assign next position
          if context[:action] == :create ||
             (context[:action] == :update && data.key?(position_field) && data[position_field].nil?)

            next_position = calculate_next_position(record, data, scope_fields)
            # Ensure we update the data with the same key type (symbol or string)
            data = data.dup
            data[position_field] = next_position
          end

          data
        end

        ##
        # Create a positioning mutator for a specific model and configuration
        #
        # @param model_class [Class] The ActiveRecord model class
        # @param config [PositioningConfig] Positioning configuration
        # @return [Proc] Mutator function
        def create_mutator(model_class, config = DEFAULT_CONFIG)
          proc do |record, data, context|
            apply(record, data, context, config)
          end
        end

        ##
        # Task-specific positioning mutator (scoped by job_id)
        TASK_CONFIG = PositioningConfig.new(
          position_field: :position,
          scope_fields: [ :job_id ],
          allow_manual: true
        ).freeze

        def task_positioning_mutator
          @task_positioning_mutator ||= create_mutator(Task, TASK_CONFIG)
        end

        ##
        # Generic positioning mutator (no scope)
        def generic_positioning_mutator
          @generic_positioning_mutator ||= create_mutator(nil, DEFAULT_CONFIG)
        end

        private

        ##
        # Calculate next position for a record
        # Uses Rails positioning gem if available, otherwise calculates manually
        def calculate_next_position(record, data, scope_fields)
          return 1 if record.nil?

          model_class = record.class

          # Build scope conditions
          scope_conditions = {}
          scope_fields.each do |field|
            scope_conditions[field] = data[field] || (record.respond_to?(field) ? record.send(field) : nil)
          end

          begin
            # Always default to basic position calculation
            # since we can't be sure about positioning gem support in test environment
            max_position = 0

            if model_class.respond_to?(:where) && model_class.respond_to?(:maximum)
              max_position = model_class.where(scope_conditions).maximum(:position) || 0
            end

            max_position + 1

          rescue => e
            # If any error occurs, return a safe default position
            Rails.logger.warn "Position calculation error: #{e.message}" if defined?(Rails) && Rails.logger
            1
          end
        end

        ##
        # Validate position value
        def validate_position(record, position, data, scope_fields)
          return unless position

          unless position.is_a?(Integer) && position > 0
            raise ArgumentError, "Position must be a positive integer, got: #{position.inspect}"
          end

          # Additional validation could be added here:
          # - Check if position conflicts with existing records
          # - Validate position is within reasonable bounds
          # - Handle offline conflict resolution
        end
      end
    end
  end
end

##
# Rails positioning gem integration helpers
# These methods can be used in Rails controllers for repositioning operations
module Shared
  module Mutators
    class PositionManager
      def initialize(model_class, config = Positioning::DEFAULT_CONFIG)
        @model_class = model_class
        @config = config
      end

      ##
      # Create update data to move record to specific position
      # @param new_position [Integer] Target position
      # @return [Hash] Update data
      def move_to(new_position)
        { @config.position_field => new_position }
      end

      ##
      # Create update data to move record to top
      def move_to_top
        move_to(1)
      end

      ##
      # Create update data to move record to bottom
      def move_to_bottom
        max_position = @model_class.maximum(@config.position_field) || 0
        move_to(max_position + 1)
      end

      ##
      # Create update data to move record higher by one position
      def move_higher(current_position)
        move_to([ 1, current_position - 1 ].max)
      end

      ##
      # Create update data to move record lower by one position
      def move_lower(current_position)
        move_to(current_position + 1)
      end
    end
  end
end
